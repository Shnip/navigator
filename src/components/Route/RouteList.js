import React from 'react';
import PropTypes from 'prop-types';
import { Droppable } from 'react-beautiful-dnd';
import Route from './Route';
import { FaBus, FaWalking, FaBicycle, FaCar } from 'react-icons/fa'
import classnames from 'classnames';

const RouteList = (props) => {
  const {routes, deleteRoute, handleMode, mode} = props;

  const handleDelete = (e) => {
    let target = e.target;
    const ul = e.currentTarget;

    while (target !== ul) {
      if (target.hasAttribute('data-delete')) {
        const id = target.dataset.delete;

        return deleteRoute(id)
      }

      target = target.parentNode;
    }
  }

  return (
    <section className="routes__list-wrapper">
      <h2 className="title routes__list-title">
        <span className="text-bold">Route</span>
        &nbsp;
        <span className="text-thin">List</span>
      </h2>

      <section className="routes__modes">
        <ul className="list route-modes" onClick={handleMode}>
          <li className={
            classnames("route-modes__mode",
                      {"route-modes__mode--active": mode === "auto"}
            )}
              data-mode="auto">
            <FaCar />
          </li>
          <li className={
            classnames("route-modes__mode",
                      {"route-modes__mode--active": mode === "bus"})}
              data-mode="bus">
            <FaBus />
          </li>
          <li className={
            classnames("route-modes__mode",
              {"route-modes__mode--active": mode === "pedestrian"})}
              data-mode="pedestrian">
            <FaWalking />
          </li>
          <li className={
            classnames("route-modes__mode",
                  {"route-modes__mode--active": mode === "bicycle"}
              )}
              data-mode="bicycle">
            <FaBicycle />
          </li>
        </ul>
      </section>

      {
        routes && routes.length > 0 ?
          <Droppable droppableId="1">
            {
              (provided, snapshot) => {
                return (
                  <ul className="list routes__list"
                      onClick={handleDelete}
                      ref={provided.innerRef}
                      style={{ backgroundColor: snapshot.isDraggingOver ? '#8bf1c4' : '' }}
                      {...provided.droppableProps}>
                    {
                      routes.map((route, index) => (
                        <Route route={route} index={index} key={route.id}/>
                      ))
                    }
                    {provided.placeholder}
                </ul>
                )
              }
            }
          </Droppable>
        :
        <span className="routes__empty">There are no routes yet</span>
      }
    </section>
  );
}

RouteList.propTypes = {
  routes: PropTypes.array,
  mode: PropTypes.string.isRequired,
  deleteRoute: PropTypes.func.isRequired,
  handleMode: PropTypes.func.isRequired,
};

export default RouteList;
