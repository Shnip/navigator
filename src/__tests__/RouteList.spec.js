import React from 'react';
import { mount } from 'enzyme';
import { DragDropContext } from 'react-beautiful-dnd';
import RouteList from '../components/Route/RouteList';


describe('test ReactCreate component', () => {
  const mockDeleteRoute = jest.fn();
  const onDragEndMock = jest.fn();
  const props = {
    routes: [],
    mode: 'auto',
    deleteRoute: mockDeleteRoute,
    handleMode: jest.fn()
  }

  const component = mount(
    <DragDropContext onDragEnd={onDragEndMock}>
      <RouteList {...props}/>
    </DragDropContext>
  );

  it('displays text when routes are empty', () => {
    expect(component.find('.routes__empty')).toHaveLength(1);
  });

  describe('when routes are not empty', () => {
    const newProps = {
      ...props,
      routes: [
        {id: 1, value: 'Moscow'},
        {id: 2, value: 'Saint-Petersburg'},
        {id: 3, value: 'Murmansk'}
      ],
      mode: 'bus'
    }

    const component = mount(
      <DragDropContext onDragEnd={onDragEndMock}>
        <RouteList {...newProps}/>
      </DragDropContext>
    );

    it('renders Routes List components', () => {
      expect(component.find('.routes__list')).toHaveLength(1);
    });

    it('renders Route components', () => {
      expect(component.find('Route')).toHaveLength(3);
    });

    it('renders Droppable component', () => {
      expect(component.find('Droppable')).toHaveLength(1);
    });

    it('calls deleteRoute when delete button is clicked', () => {
      component.find('[data-delete]').at(0).simulate('click', {
        target: {
          hasAttribute: () => true,
          dataset: {
            delete: 1
          }
        }
      });

      expect(mockDeleteRoute).toHaveBeenCalledTimes(1);
    });

    test ('bus must have an active class', () => {
      expect(component.find("[data-mode='bus']").hasClass('route-modes__mode--active')).toBeTruthy()
    });

    it ('renders properly', () => {
      expect(component).toMatchSnapshot();
    });
  })
})
