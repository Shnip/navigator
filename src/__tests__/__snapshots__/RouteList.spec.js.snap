// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test ReactCreate component when routes are not empty renders properly 1`] = `ReactWrapper {}`;
